### PR TITLE
Recorder value refactor

### PR DESCRIPTION
**Pull request by jetuk**
Here's the work I've done to `Recorders` recently. 

This currently hangs off the virtual storage branch due to my disorganisation.

### CODE REVIEW

File: pywr/_recorders.pyx [hunk 1]
```diff
@@ -1,20 +1,44 @@
...
 cimport numpy as np
-
 from pywr._core cimport Timestep
-
 import pandas as pd
+from past.builtins import basestring
 
 recorder_registry = set()
 
+cdef enum AggFuncs:
```
Comment (snorfalorpagus) [inline pywr/_recorders.pyx]: We should move this (and the one in `_parameters`) into `_core`.

**Comment by jetuk** [comment]
Should I rebase this to master without the virtual storage stuff?

**Comment by snorfalorpagus** [comment]
Yes please. Probably easiest to cherry pick the commits:

```
git checkout master
git checkout -b new_branch
git cherry-pick xxxxxxx
git cherry-pick xxxxxxx
git cherry-pick xxxxxxx
```

Where xxx are the commit IDs you want.

File: pywr/_recorders.pyx (hunk 1 above)
Comment (jetuk) [inline pywr/_recorders.pyx]: I think I hit this trying to do this ...

https://github.com/cython/cython/issues/531

https://groups.google.com/forum/#!msg/cython-users/ZoLsLHwnUY4/74_JZaZvAAAJ

Comment (snorfalorpagus) [inline pywr/_recorders.pyx]: I don't think the enums need to be cpdef. They're just used so the compiler can optimise the switch/case statement. The end-user is expected to use strings (e.g. "sum").

**Comment by snorfalorpagus** [comment]
I think for now I'd be OK with you just keeping `AggFuncs` in `_recorders` (with a duplicate in `_parameter`) and renaming the constants to use upper case to avoid collision with builtin names.

**Comment by jetuk** [comment]
I made the constant changes and merged this.

**Pull request merged**